src/core/thread/network_data_notifier.cpp

```diff
@@ -152,6 +166,38 @@ void Notifier::HandleCoapResponse(Error aResult)
     }
 }
 
+#if OPENTHREAD_FTD && OPENTHREAD_CONFIG_BORDER_ROUTER_ENABLE && OPENTHREAD_CONFIG_BORDER_ROUTER_REQUEST_ROUTER_ROLE
+void Notifier::RequestRouterRoleIfEligible(void)
+{
+    VerifyOrExit(Get<Mle::MleRouter>().IsChild() && Get<Mle::MleRouter>().IsRouterEligible());
+
+    // RouterUpgradeThreshold can be explicitly set to zero in some of
+    // cert tests to disallow device to become router.
+
+    VerifyOrExit(Get<Mle::MleRouter>().GetRouterUpgradeThreshold() != 0);
+
+    // Check the local network data to ensure that we are a border
+    // router providing IP connectivity and therefore eligible to
+    // request router role upgrade with `kBorderRouterRequst`
+    // status.
+
+    VerifyOrExit(Get<Local>().CountBorderRouters(kAnyRole) > 0);
+
+    if (Get<RouterTable>().GetActiveRouterCount() >= Get<Mle::MleRouter>().GetRouterUpgradeThreshold())
+    {
+        VerifyOrExit(Get<Leader>().CountBorderRouters(kRouterRoleOnly) <
+                     Mle::kRouterUpgradeBorderRouterRequestThreshold);
+
+        LogInfo("Requesting router role as BR when above upgrade threshold");
+    }
+
+    IgnoreError(Get<Mle::MleRouter>().BecomeRouter(ThreadStatusTlv::kBorderRouterRequst));
```

Comment:
This line is outside the brackets of above 'if' statement - I think it should be inside. I.e. only request upgrade with this reason if  ActiveRouterCount >= UpgradeThreshold.

Reply:
That makes sense. If the number of routers is less than router upgrade threshold, then this BR will attempt to upgrade to router using existing Thread 1.x mechanisms.

I think the more important consideration here is whether we want to apply some random delay. If there are a large number of BRs in the network, I could see a large number of requests being sent to the Leader at the same time whenever the network data is updated and the conditions to upgrade are satisfied.

Reply:
Thanks guys.

My idea is that as soon as we configured as an eligible BR:
- If number of active routers is below threshold, we don't wait for normal mechanism to upgrade to a router role and request it immediately (this gives more chance for BRs to start as router from beginning).
  - We can still request it with `kBorderRouterRequst` reason and leader is expected to treat as a normal request when we are below the active routers threshold.
- If number of active is above threshold, then we would count the number of existing BRs (from network data) and decide based on that whether to attempt to become router or not.

I understand it going somewhat beyond what we described in the JIRA (though it still using the new mechanism) but I feel this may be useful behavior.

Thoughts?

> I think the more important consideration here is whether we want to apply some random delay. If there are a large number of BRs in the network, I could see a large number of requests being sent to the Leader at the same time whenever the network data is updated and the conditions to upgrade are satisfied.

On adding some delay to send the the requests, I think that make sense. We may be able to re-use the timer in this module (probably need to maintain some sort of state). Will try to add this later in next push.

Reply:
Thinking more about this, decided to keep it simple so we use the new `kBorderRouterRequest` only when we are above the active router threshold (i.e., when we are below the threshold, the normal mechanism for router upgrade is used).

Added this in the new push. Also there is a randomly chosen delay (up to 10 second) before sending "Address Solicit" with this reason once device becomes eligible for such a request.

Reply:
> Thinking more about this, decided to keep it simple so we use the new kBorderRouterRequest only when we are above the active router threshold (i.e., when we are below the threshold, the normal mechanism for router upgrade is used).

Agree. Use normal mechanism whenever possible is more conservative, thus less risky.